fdbserver/DDTxnProcessor.actor.cpp

```diff
@@ -538,3 +538,168 @@ Future<HealthMetrics> DDTxnProcessor::getHealthMetrics(bool detailed) const {
 Future<Optional<Value>> DDTxnProcessor::readRebalanceDDIgnoreKey() const {
 	return DDTxnProcessorImpl::readRebalanceDDIgnoreKey(cx);
 }
+
+Future<std::vector<std::pair<StorageServerInterface, ProcessClass>>>
+DDMockTxnProcessor::getServerListAndProcessClasses() {
+	std::vector<std::pair<StorageServerInterface, ProcessClass>> res;
+	for (auto& [_, mss] : mgs->allServers) {
+		res.emplace_back(mss.ssi, ProcessClass(ProcessClass::StorageClass, ProcessClass::DBSource));
+	}
+	return res;
+}
+
+std::pair<std::set<std::vector<UID>>, std::set<std::vector<UID>>> getAllTeamsInRegion(
+    const std::vector<DDShardInfo>& shards) {
+	std::set<std::vector<UID>> primary, remote;
+	for (auto& info : shards) {
+		if (!info.primarySrc.empty())
+			primary.emplace(info.primarySrc);
+		if (!info.primaryDest.empty())
+			primary.emplace(info.primaryDest);
+		if (!info.remoteSrc.empty())
+			remote.emplace(info.remoteSrc);
+		if (!info.remoteDest.empty())
+			remote.emplace(info.remoteDest);
+	}
+	return { primary, remote };
+}
+
+inline void transformTeamsToServerIds(std::vector<ShardsAffectedByTeamFailure::Team>& teams,
+                                      std::vector<UID>& primaryIds,
+                                      std::vector<UID>& remoteIds) {
+	std::set<UID> primary, remote;
+	for (auto& team : teams) {
+		team.primary ? primary.insert(team.servers.begin(), team.servers.end())
+		             : remote.insert(team.servers.begin(), team.servers.end());
+	}
+	primaryIds = std::vector<UID>(primary.begin(), primary.end());
+	remoteIds = std::vector<UID>(remote.begin(), remote.end());
+}
+
+// reconstruct DDShardInfos from shardMapping
+std::vector<DDShardInfo> DDMockTxnProcessor::getDDShardInfos() const {
+	std::vector<DDShardInfo> res;
+	res.reserve(mgs->shardMapping->getNumberOfShards());
+	auto allRange = mgs->shardMapping->getAllRanges();
+	ASSERT(allRange.end().begin() == allKeys.end);
+	for (auto it = allRange.begin(); it != allRange.end(); ++it) {
+		// FIXME: now just use anonymousShardId
+		KeyRangeRef curRange = it->range();
+		DDShardInfo info(curRange.begin);
+
+		auto teams = mgs->shardMapping->getTeamsFor(curRange);
+		if (!teams.first.empty() && !teams.second.empty()) {
+			CODE_PROBE(true, "Mock InitialDataDistribution In-Flight shard");
+			info.hasDest = true;
+			info.destId = anonymousShardId;
+			info.srcId = anonymousShardId;
+			transformTeamsToServerIds(teams.second, info.primarySrc, info.remoteSrc);
+			transformTeamsToServerIds(teams.first, info.primaryDest, info.remoteDest);
+		} else if (!teams.first.empty()) {
+			CODE_PROBE(true, "Mock InitialDataDistribution Static shard");
+			info.srcId = anonymousShardId;
+			transformTeamsToServerIds(teams.first, info.primarySrc, info.remoteSrc);
+		} else {
+			ASSERT(false);
```

Comment:
I've sometimes found it useful to say something like ASSERT(!"Shard has not associated servers"); so the assertion failure has a clearer signature.